core/src/main/scala/akka/kafka/KafkaConsumerActor.scala

```diff
@@ -0,0 +1,205 @@
+          if (exception != null) reply ! Status.Failure(exception)
+          else reply ! Committed(offsets.asScala.toMap)
+        }
+      })
+      //right now we can not store commits in consumer - https://issues.apache.org/jira/browse/KAFKA-3412
+      pollExpected = true
+      poll()
+    case Subscribe(topics, listener) =>
+      consumer.subscribe(topics.toList.asJava, listener)
+    case SubscribePattern(pattern, listener) =>
+      consumer.subscribe(Pattern.compile(pattern), listener)
+    case Poll =>
+      pollExpected = true
+      poll()
+    case RequestMessages(topics) =>
+      requests ++= topics.map(_ -> sender()).toMap
+      pollExpected = true
+      poll()
+    case Stop =>
+      if (commitsInProgress == 0) {
+        context.stop(self)
+      }
+      else {
+        stopInProgress = true
+        context.become(stopping)
+      }
+  }
+
+  def stopping: Receive = LoggingReceive {
+    case Watch => sender() ! finished.future
+    case Assign(tps) =>
+    case AssignWithOffset(tps) =>
+    case Commit(offsets) =>
+      sender() ! Status.Failure(StoppingException())
+    case Subscribe(topics, listener) =>
+    case SubscribePattern(pattern, listener) =>
+    case Poll =>
+      pollExpected = true
+      poll()
+    case RequestMessages(topics) =>
+      sender() ! Status.Failure(StoppingException())
+    case Stop =>
+  }
+
+  override def preStart(): Unit = {
+    super.preStart()
+    requests = Map.empty[TopicPartition, ActorRef]
+    consumer = settings.createKafkaConsumer()
+    nextScheduledPoll = None
+    commitsInProgress = 0
+    pollExpected = false
+    finished = Promise()
+    schedulePoll()
+  }
+
+  override def postStop(): Unit = {
+    nextScheduledPoll.foreach(_.cancel())
+    consumer.close()
+    finished.trySuccess(Done)
+    super.postStop()
+  }
+
+  def poll() = {
+    if (pollExpected) {
+      pollExpected = false
+      nextScheduledPoll.foreach(_.cancel())
+      nextScheduledPoll = None
+
+      //set partitions to fetch
+      val partitionsToFetch = requests.keys.toSet
+      consumer.assignment().asScala.foreach { tp =>
+        if (partitionsToFetch.contains(tp)) consumer.resume(tp)
+        else consumer.pause(tp)
+      }
+
+      val rawResult = consumer.poll(pollTimeout().toMillis)
+      if (!rawResult.isEmpty) {
+        // split tps by reply actor
+        val replyByTP = requests
+          .groupBy { case (tp, ref) => ref }
+          .mapValues(_.keys.toSet)
+
+        //send messages to actors
+        replyByTP.foreach {
+          case (ref, tps) =>
+            //gather all messages for ref
+            val messages = tps.foldLeft[Iterator[ConsumerRecord[K, V]]](Iterator.empty) {
+              case (acc, tp) =>
+                val tpMessages = rawResult.records(tp).asScala.iterator
+                if (acc.isEmpty) tpMessages
+                else acc ++ tpMessages
+            }
+            if (messages.nonEmpty) {
+              ref ! Messages(messages)
```

Comment:
IMO it is pretty safe. 
Here it is collection copying in kafka - https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java#L46
Also I create fresh iterator wrapper every time and only one consumer uses it.